File: test/TestStrategyGrid/GridStrategy.cpp
#include "GridStrategy.h"
#include <Spark/Core/Logger/Logger.h>
#include <limits>

using namespace spark::core;

namespace quanttrading::teststrategygrid
{
GridStrategy::GridStrategy(quanttrading::BackTestApi* backTestApi, const char* accountID, const GridParams& gridParams)
	:StrategyBase(backTestApi, accountID), m_Params(gridParams)
{
	m_Slots.resize(gridParams.GridCount * 2);
	for (int level = 0; level < gridParams.GridCount; ++level)
	{
		m_Slots[level].Direction = DirectionType::Buy;
		m_Slots[gridParams.GridCount + level].Direction = DirectionType::Sell;
	}
}

void GridStrategy::OnStart()
{
	SubscribeTick(m_Params.ExchangeID.c_str(), m_Params.InstrumentID.c_str());
}

void GridStrategy::OnTick(const DepthMarketDataField* depthMarketData)
{
	if (!m_AwaitingAnchor)
	{
		return;
	}
	constexpr PriceType InvalidPrice = std::numeric_limits<PriceType>::max();
	PriceType anchorPrice = depthMarketData->LastPrice;
	if (anchorPrice <= 0 || anchorPrice == InvalidPrice)
	{
		return;
	}
	m_AwaitingAnchor = false;
	WriteLog(LogLevel::Info, "Anchor price: %f", anchorPrice);
	PlaceLadder(anchorPrice);
}

void GridStrategy::PlaceLadder(PriceType anchorPrice)
{
	for (int level = 0; level < m_Params.GridCount; ++level)
	{
		GridSlot& buySlot = m_Slots[level];
		if (buySlot.State == GridSlotState::Empty)
		{
			buySlot.OpenPrice = anchorPrice - m_Params.GridStep * (level + 1);
			PlaceOpenOrder(buySlot);
		}
		GridSlot& sellSlot = m_Slots[m_Params.GridCount + level];
		if (sellSlot.State == GridSlotState::Empty)
		{
			sellSlot.OpenPrice = anchorPrice + m_Params.GridStep * (level + 1);
			PlaceOpenOrder(sellSlot);
		}
	}
}

void GridStrategy::PlaceOpenOrder(GridSlot& gridSlot)
{
	gridSlot.OpenFilledVolume = 0;
	if (gridSlot.Direction == DirectionType::Buy)
	{
		gridSlot.OpenClientOrderID = BuyOpen(m_Params.ExchangeID.c_str(), m_Params.InstrumentID.c_str(), gridSlot.OpenPrice, m_Params.VolumePerGrid);
	}
	else
	{
		gridSlot.OpenClientOrderID = SellOpen(m_Params.ExchangeID.c_str(), m_Params.InstrumentID.c_str(), gridSlot.OpenPrice, m_Params.VolumePerGrid);
	}
	gridSlot.State = GridSlotState::OpenPending;
	WriteLog(LogLevel::Info, "Place open order, ClientOrderID:%d Direction:%d Price:%f", gridSlot.OpenClientOrderID, (int)gridSlot.Direction, gridSlot.OpenPrice);
}

void GridStrategy::PlaceCloseOrder(GridSlot& gridSlot, VolumeType volume)
{
	if (gridSlot.Direction == DirectionType::Buy)
	{
		gridSlot.CloseClientOrderID = SellClose(m_Params.ExchangeID.c_str(), m_Params.InstrumentID.c_str(), gridSlot.ClosePrice, volume);
	}
	else
	{
		gridSlot.CloseClientOrderID = BuyClose(m_Params.ExchangeID.c_str(), m_Params.InstrumentID.c_str(), gridSlot.ClosePrice, volume);
	}
	gridSlot.State = GridSlotState::ClosePending;
	WriteLog(LogLevel::Info, "Place close order, ClientOrderID:%d Price:%f Volume:%lld", gridSlot.CloseClientOrderID, gridSlot.ClosePrice, volume);
}

void GridStrategy::OnTrade(const TradeField* trade, ClientOrderIDType clientOrderID)
{
	if (clientOrderID == 0)
	{
		return;
	}
	if (GridSlot* gridSlot = FindSlotByOpenOrder(clientOrderID))
	{
		HandleOpenTrade(trade, gridSlot);
	}
	else if (GridSlot* gridSlot = FindSlotByCloseOrder(clientOrderID))
	{
		HandleCloseTrade(trade, gridSlot);
	}
}

void GridStrategy::OnOrder(const OrderField* order)
{
	if (order->OrderStatus != OrderStatusType::Canceled && order->OrderStatus != OrderStatusType::PartTradedCanceled)
	{
		return;
	}
	if (GridSlot* gridSlot = FindSlotByOpenOrder(order->ClientOrderID))
	{
		HandleOpenOrderCanceled(gridSlot);
	}
	else if (GridSlot* gridSlot = FindSlotByCloseOrder(order->ClientOrderID))
	{
		HandleCloseOrderCanceled(gridSlot);
	}
}

void GridStrategy::HandleOpenTrade(const TradeField* trade, GridSlot* gridSlot)
{
	if (gridSlot->State != GridSlotState::OpenPending)
	{
		return;
	}
	gridSlot->OpenFillPrice = trade->Price;
	gridSlot->OpenFilledVolume += trade->Volume;
	if (gridSlot->OpenFilledVolume < m_Params.VolumePerGrid)
	{
		return;
	}
	UpdateClosePrice(*gridSlot);
	PlaceCloseOrder(*gridSlot, gridSlot->OpenFilledVolume);
}

void GridStrategy::HandleCloseTrade(const TradeField* trade, GridSlot* gridSlot)
{
	if (gridSlot->State != GridSlotState::ClosePending)
	{
		return;
	}
	gridSlot->CloseFilledVolume += trade->Volume;
	if (gridSlot->CloseFilledVolume < gridSlot->OpenFilledVolume)
	{
		return;
	}
	double profitPerUnit = 0.0;
	if (gridSlot->Direction == DirectionType::Buy)
	{
		profitPerUnit = trade->Price - gridSlot->OpenFillPrice;
	}
	else
	{
		profitPerUnit = gridSlot->OpenFillPrice - trade->Price;
	}
	double pairProfit = profitPerUnit * trade->Volume * trade->VolumeMultiple - trade->Commission;
	m_RealizedProfit += pairProfit;
	m_TotalCommission += trade->Commission;
	++m_ClosedPairCount;
	gridSlot->State = GridSlotState::Closed;
	WriteLog(LogLevel::Info, "Pair closed, open:%f close:%f profit:%f", gridSlot->OpenFillPrice, trade->Price, pairProfit);
}

void GridStrategy::UpdateClosePrice(GridSlot& gridSlot)
{
	if (gridSlot.Direction == DirectionType::Buy)
	{
		gridSlot.ClosePrice = gridSlot.OpenFillPrice + m_Params.GridStep;
	}
	else
	{
		gridSlot.ClosePrice = gridSlot.OpenFillPrice - m_Params.GridStep;
	}
}

void GridStrategy::ResetSlotToEmpty(GridSlot& gridSlot)
{
	// 清净周期态字段：格位跨周期复用，残留的 CloseFilledVolume 会使新周期的平仓剩余量
	// 计算为 0，撤单后不再补平仓单，格位带持仓困死在 ClosePending
	gridSlot.State = GridSlotState::Empty;
	gridSlot.OpenClientOrderID = 0;
	gridSlot.CloseClientOrderID = 0;
	gridSlot.OpenFilledVolume = 0;
	gridSlot.CloseFilledVolume = 0;
}

void GridStrategy::HandleOpenOrderCanceled(GridSlot* gridSlot)
{
	if (gridSlot->State != GridSlotState::OpenPending)
	{
		return;
	}
	if (gridSlot->OpenFilledVolume == 0)
	{
		ResetSlotToEmpty(*gridSlot);
		WriteLog(LogLevel::Info, "Open order canceled without fill, slot reset to Empty, Price:%f", gridSlot->OpenPrice);
		return;
	}
	// 部分成交后被日终撤销：按已成交量即时补平仓单，新单经引擎队列在次一交易日撮合
	UpdateClosePrice(*gridSlot);
	PlaceCloseOrder(*gridSlot, gridSlot->OpenFilledVolume);
	WriteLog(LogLevel::Warning, "Partially filled open order canceled by day-end settlement, close volume:%lld", gridSlot->OpenFilledVolume);
}

void GridStrategy::HandleCloseOrderCanceled(GridSlot* gridSlot)
{
	if (gridSlot->State != GridSlotState::ClosePending)
	{
		return;
	}
	VolumeType remainingVolume = gridSlot->OpenFilledVolume - gridSlot->CloseFilledVolume;
	if (remainingVolume <= 0)
	{
		WriteLog(LogLevel::Warning, "Close order canceled with zero remaining volume, ClosePrice:%f", gridSlot->ClosePrice);
		return;
	}
	// 平仓单被日终撤销：按剩余未平量以原价位重下平仓单，CloseFilledVolume 累计保留
	PlaceCloseOrder(*gridSlot, remainingVolume);
	WriteLog(LogLevel::Warning, "Close order canceled by day-end settlement, re-place close volume:%lld", remainingVolume);
}

void GridStrategy::OnInsertOrderRsp(const ReqInsertOrderField* reqInsertOrder, const RspInfoField* rspInfo)
{
	if (reqInsertOrder == nullptr || rspInfo == nullptr || rspInfo->ErrorID == 0)
	{
		return;
	}
	ClientOrderIDType clientOrderID = reqInsertOrder->ClientOrderID;
	if (GridSlot* gridSlot = FindSlotByOpenOrder(clientOrderID))
	{
		if (gridSlot->State == GridSlotState::OpenPending)
		{
			ResetSlotToEmpty(*gridSlot);
			WriteLog(LogLevel::Warning, "Open order rejected, slot reset to Empty, Price:%f", gridSlot->OpenPrice);
		}
	}
	else if (GridSlot* gridSlot = FindSlotByCloseOrder(clientOrderID))
	{
		// 平仓单被拒：保持 OpenFilled（持仓留存），由 OnEnd 总结提示剩余持仓
		WriteLog(LogLevel::Error, "Close order rejected, position remains, ClosePrice:%f", gridSlot->ClosePrice);
	}
}

void GridStrategy::OnSessionBegin(const SessionBeginField* sessionBegin)
{
	// 引擎已在结算时统一撤单（撤单回报先于 SessionEnd）：OpenPending 格位已经撤单回报复位 Empty
	// 或转换为 ClosePending，此处仅需复位 Closed 格等待重锚
	for (auto& gridSlot : m_Slots)
	{
		if (gridSlot.State == GridSlotState::Closed)
		{
			ResetSlotToEmpty(gridSlot);
		}
	}
	m_AwaitingAnchor = true;
}

void GridStrategy::OnSessionEnd(const SessionEndField* sessionEnd)
{
	// 引擎日切结算已撤销全部未成交挂单（撤单回报先于本回调）：OpenPending 格位已在 OnOrder 中
	// 复位/转换，此处统计的是转换后仍带平仓单、次一交易日继续工作的格位
	int closePendingCount = 0;
	for (auto& gridSlot : m_Slots)
	{
		if (gridSlot.State == GridSlotState::ClosePending)
		{
			++closePendingCount;
		}
	}
	WriteLog(LogLevel::Info, "Session end: %d slots closing next day after day-end cancel sweeps", closePendingCount);
}

void GridStrategy::OnEnd()
{
	WriteLog(LogLevel::Info, "Grid strategy end: closedPairs:%d realizedProfit:%f totalCommission:%f longPosition:%lld shortPosition:%lld",
		m_ClosedPairCount, m_RealizedProfit, m_TotalCommission,
		GetLongPosition(m_Params.InstrumentID.c_str()), GetShortPosition(m_Params.InstrumentID.c_str()));
}

GridStrategy::GridSlot* GridStrategy::FindSlotByOpenOrder(ClientOrderIDType clientOrderID)
{
	for (auto& gridSlot : m_Slots)
	{
		if (gridSlot.OpenClientOrderID == clientOrderID)
		{
			return &gridSlot;
		}
	}
	return nullptr;
}
GridStrategy::GridSlot* GridStrategy::FindSlotByCloseOrder(ClientOrderIDType clientOrderID)
{
	for (auto& gridSlot : m_Slots)
	{
		if (gridSlot.CloseClientOrderID == clientOrderID)
		{
			return &gridSlot;
		}
	}
	return nullptr;
}
}

File: test/PythonStrategyGrid/grid_strategy.py
"""成对网格策略 Python 移植（对应 test/TestStrategyGrid/GridStrategy.cpp，经 QuantTrading.pyd 驱动回测引擎）。

运行方式（脚本与配置由构建拷贝至 bin/Release，引擎从 CWD 读 BackTest.json 与 TestStrategyGrid.json）：
    cd bin/Release && python grid_strategy.py
"""

import json
import os
import sys
from dataclasses import dataclass
from enum import IntEnum

REPO_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(REPO_ROOT, "bin", "Release"))

import QuantTrading as qt


class GridSlotState(IntEnum):
    Empty = 0
    OpenPending = 1
    OpenFilled = 2
    ClosePending = 3
    Closed = 4


class GridSlot:
    def __init__(self, direction):
        self.direction = direction
        self.state = GridSlotState.Empty
        self.open_price = 0.0
        self.open_fill_price = 0.0
        self.close_price = 0.0
        self.open_client_order_id = 0
        self.close_client_order_id = 0
        self.open_filled_volume = 0
        self.close_filled_volume = 0


@dataclass
class GridParams:
    grid_step: float
    grid_count: int
    volume_per_grid: int
    exchange_id: str
    instrument_id: str
    bar_preces: str = ""   # 策略期望 bar 周期（"5m" 格式，同 BackTest.json BarPreces）；空=on_bar 纯透传


class GridStrategy(qt.StrategyBase):
    """格位状态机 Empty → OpenPending → OpenFilled → ClosePending → Closed，日级重锚。

    [0, GridCount) 为买开格，[GridCount, 2×GridCount) 为卖开格；平仓价取开仓成交价 ∓ 步长，与锚点无关。
    引擎日切结算统一撤销全部未成交挂单：零成交开仓格复位 Empty 等次日重锚重挂；部分成交开仓格
    与被撤平仓格按已成交/剩余量即时补平仓单，新平仓单经引擎队列在次一交易日撮合。
    """

    def __init__(self, backtest_api, account_id, params):
        super().__init__(backtest_api=backtest_api, account_id=account_id)
        self.params = params
        if params.bar_preces:
            self.declare_bar_period(params.bar_preces)
        self.slots = [GridSlot(qt.DirectionType.Buy) for _ in range(params.grid_count)] + \
            [GridSlot(qt.DirectionType.Sell) for _ in range(params.grid_count)]
        self.awaiting_anchor = True
        self.closed_pair_count = 0
        self.realized_profit = 0.0
        self.total_commission = 0.0

    def on_start(self):
        self.subscribe_tick(self.params.exchange_id, self.params.instrument_id)

    def on_tick(self, market_data):
        if not self.awaiting_anchor:
            return
        anchor_price = market_data.last_price
        if anchor_price <= 0 or anchor_price == sys.float_info.max:
            return
        self.awaiting_anchor = False
        print("Anchor price: %f" % anchor_price)
        self.place_ladder(anchor_price)

    def place_ladder(self, anchor_price):
        for level in range(self.params.grid_count):
            buy_slot = self.slots[level]
            if buy_slot.state == GridSlotState.Empty:
                buy_slot.open_price = anchor_price - self.params.grid_step * (level + 1)
                self.place_open_order(buy_slot)
            sell_slot = self.slots[self.params.grid_count + level]
            if sell_slot.state == GridSlotState.Empty:
                sell_slot.open_price = anchor_price + self.params.grid_step * (level + 1)
                self.place_open_order(sell_slot)

    def place_open_order(self, slot):
        slot.open_filled_volume = 0
        if slot.direction == qt.DirectionType.Buy:
            slot.open_client_order_id = self.buy_open(
                self.params.exchange_id, self.params.instrument_id, slot.open_price, self.params.volume_per_grid)
        else:
            slot.open_client_order_id = self.sell_open(
                self.params.exchange_id, self.params.instrument_id, slot.open_price, self.params.volume_per_grid)
        slot.state = GridSlotState.OpenPending

    def place_close_order(self, slot, volume):
        if slot.direction == qt.DirectionType.Buy:
            slot.close_client_order_id = self.sell_close(
                self.params.exchange_id, self.params.instrument_id, slot.close_price, volume)
        else:
            slot.close_client_order_id = self.buy_close(
                self.params.exchange_id, self.params.instrument_id, slot.close_price, volume)
        slot.state = GridSlotState.ClosePending

    def on_trade(self, trade, client_order_id):
        if client_order_id == 0:
            return
        slot = self.find_slot_by_open_order(client_order_id)
        if slot is not None:
            self.handle_open_trade(trade, slot)
            return
        slot = self.find_slot_by_close_order(client_order_id)
        if slot is not None:
            self.handle_close_trade(trade, slot)

    def on_order(self, order):
        if order.order_status != qt.OrderStatusType.Canceled and \
                order.order_status != qt.OrderStatusType.PartTradedCanceled:
            return
        slot = self.find_slot_by_open_order(order.client_order_id)
        if slot is not None:
            self.handle_open_order_canceled(slot)
            return
        slot = self.find_slot_by_close_order(order.client_order_id)
        if slot is not None:
            self.handle_close_order_canceled(slot)

    def on_insert_order_rsp(self, req_insert_order, rsp_info):
        if rsp_info.error_id == 0:
            return
        slot = self.find_slot_by_open_order(req_insert_order.client_order_id)
        if slot is not None:
            if slot.state == GridSlotState.OpenPending:
                self.reset_slot_to_empty(slot)
                print("Open order rejected, slot reset to Empty, Price:%f" % slot.open_price)
            return
        slot = self.find_slot_by_close_order(req_insert_order.client_order_id)
        if slot is not None:
            print("Close order rejected, position remains, ClosePrice:%f" % slot.close_price)

    def handle_open_trade(self, trade, slot):
        if slot.state != GridSlotState.OpenPending:
            return
        slot.open_fill_price = trade.price
        slot.open_filled_volume += trade.volume
        if slot.open_filled_volume < self.params.volume_per_grid:
            return
        self.update_close_price(slot)
        self.place_close_order(slot, slot.open_filled_volume)

    def handle_close_trade(self, trade, slot):
        if slot.state != GridSlotState.ClosePending:
            return
        slot.close_filled_volume += trade.volume
        if slot.close_filled_volume < slot.open_filled_volume:
            return
        if slot.direction == qt.DirectionType.Buy:
            profit_per_unit = trade.price - slot.open_fill_price
        else:
            profit_per_unit = slot.open_fill_price - trade.price
        pair_profit = profit_per_unit * trade.volume * trade.volume_multiple - trade.commission
        self.realized_profit += pair_profit
        self.total_commission += trade.commission
        self.closed_pair_count += 1
        slot.state = GridSlotState.Closed
        print("Pair closed, open:%f close:%f profit:%f" % (slot.open_fill_price, trade.price, pair_profit))

    def update_close_price(self, slot):
        if slot.direction == qt.DirectionType.Buy:
            slot.close_price = slot.open_fill_price + self.params.grid_step
        else:
            slot.close_price = slot.open_fill_price - self.params.grid_step

    def reset_slot_to_empty(self, slot):
        # 清净周期态字段：格位跨周期复用，残留 CloseFilledVolume 会使平仓剩余量算成 0，格位困死
        slot.state = GridSlotState.Empty
        slot.open_client_order_id = 0
        slot.close_client_order_id = 0
        slot.open_filled_volume = 0
        slot.close_filled_volume = 0

    def handle_open_order_canceled(self, slot):
        if slot.state != GridSlotState.OpenPending:
            return
        if slot.open_filled_volume == 0:
            self.reset_slot_to_empty(slot)
            print("Open order canceled without fill, slot reset to Empty, Price:%f" % slot.open_price)
            return
        self.update_close_price(slot)
        self.place_close_order(slot, slot.open_filled_volume)
        print("Partially filled open order canceled by day-end settlement, close volume:%d" % slot.open_filled_volume)

    def handle_close_order_canceled(self, slot):
        if slot.state != GridSlotState.ClosePending:
            return
        remaining_volume = slot.open_filled_volume - slot.close_filled_volume
        if remaining_volume <= 0:
            print("Close order canceled with zero remaining volume, ClosePrice:%f" % slot.close_price)
            return
        self.place_close_order(slot, remaining_volume)
        print("Close order canceled by day-end settlement, re-place close volume:%d" % remaining_volume)

    def on_session_begin(self, session_begin):
        # 引擎已在结算时统一撤单（撤单回报先于 SessionEnd）：OpenPending 格位已复位/转换，
        # 此处仅需复位 Closed 格等待重锚
        for slot in self.slots:
            if slot.state == GridSlotState.Closed:
                self.reset_slot_to_empty(slot)
        self.awaiting_anchor = True

    def on_session_end(self, session_end):
        close_pending_count = 0
        for slot in self.slots:
            if slot.state == GridSlotState.ClosePending:
                close_pending_count += 1
        print("Session end: %d slots closing next day after day-end cancel sweeps" % close_pending_count)

    def on_end(self):
        print("Grid strategy end: closedPairs:%d realizedProfit:%f totalCommission:%f longPosition:%d shortPosition:%d" % (
            self.closed_pair_count, self.realized_profit, self.total_commission,
            self.get_long_position(self.params.instrument_id), self.get_short_position(self.params.instrument_id)))

    def find_slot_by_open_order(self, client_order_id):
        for slot in self.slots:
            if slot.open_client_order_id == client_order_id:
                return slot
        return None

    def find_slot_by_close_order(self, client_order_id):
        for slot in self.slots:
            if slot.close_client_order_id == client_order_id:
                return slot
        return None


RESULT_FILE_NAME = "result.json"

# 退出码契约（与 src/BackTest/RunResult.h 同源）：
# 0 成功 / 1 宿主启动失败 / 2 结果文件缺失或不可解析 / 3 引擎报告失败
EXIT_CODE_HOST_INIT_FAILED = 1
EXIT_CODE_RESULT_FILE_UNREADABLE = 2
EXIT_CODE_ENGINE_FAILED = 3


def read_exit_code(result_file_path):
    """按结果契约把 result.json 映射成退出码；文件缺失或坏掉一律算「本轮没走完收尾」。"""
    try:
        with open(result_file_path, encoding="utf-8") as result_file:
            run_result = json.load(result_file)
    except (OSError, ValueError):
        return EXIT_CODE_RESULT_FILE_UNREADABLE
    return 0 if run_result.get("Success", False) else EXIT_CODE_ENGINE_FAILED


def main():
    with open("TestStrategyGrid.json", encoding="utf-8") as config_file:
        config = json.load(config_file)
    qt.init_logger(sys.argv[0], config["LogLevel"])
    # 启动前先删陈旧结果，让「文件不存在」等价于「本轮没走完收尾」
    try:
        os.remove(RESULT_FILE_NAME)
    except FileNotFoundError:
        pass
    api = qt.create_backtest_api()
    params = GridParams(
        grid_step=config["GridStep"],
        grid_count=config["GridCount"],
        volume_per_grid=config["VolumePerGrid"],
        exchange_id=config["ExchangeId"],
        instrument_id=config["InstrumentId"],
        bar_preces=config.get("BarPreces", ""))
    strategy = GridStrategy(api, config["AccountId"], params)
    if not strategy.start():
        qt.shutdown_logger()
        return EXIT_CODE_HOST_INIT_FAILED
    strategy.wait_for_end()
    qt.shutdown_logger()
    return read_exit_code(RESULT_FILE_NAME)


if __name__ == "__main__":
    sys.exit(main())
